packages/chain/statemgr/syncmgr.go

```diff
@@ -79,8 +79,17 @@ func (sm *stateManager) doSyncActionIfNeeded() {
 		sm.domain.SetFallbackMode(true)
 	}
 	for i := startSyncFromIndex; i <= sm.stateOutput.GetStateIndex(); i++ {
+		if sm.wal.IsSynced(i) {
+			return
+		}
 		requestBlockRetryTime := sm.syncingBlocks.getRequestBlockRetryTime(i)
 		blockCandidatesCount := sm.syncingBlocks.getBlockCandidatesCount(i)
+		if blockCandidatesCount == 0 {
+			if blockAdded, blockValidated := sm.candidateBlockInWAL(i); blockAdded && blockValidated {
+				blockCandidatesCount++
+				sm.wal.Synced(i)
+			}
```

Comment:
Now there is no need to check if the new candidate block is approved. `getApprovedBlockCandidatesCount` will do that.
Moreover, I see no need to enrich Wal interface with the things, that are needed only by state manager. State manager should keep its internals for itself.
```suggestion
			if sm.candidateBlockInWAL(i) { // blockAdded
				blockCandidatesCount++
				sm.syncingBlocks.setObtainedFromWal(i)
			}
```
`setObtainedFromWal` must be implemented. I think, this can set a boolean field in i-th `syncingBlock` structure of `syncingBlocks.blocks`.